### PR TITLE
2107 idf plot selection

### PR DESCRIPTION
**Pull request by lucas-wilkins**
Goal: add the IDF plot as a tab

* Added IDF tab to corfunc
* Further refactored corfunc plotting
* Added IDF to reports

### CODE REVIEW

**Review by wpotrzebowski — COMMENTED** [review]
I have problems running it from the installer on Mac. 

Traceback (most recent call last):
  File "sasview.py", line 15, in <module>
  File "sas/qtgui/MainWindow/MainWindow.py", line 113, in run_sasview
  File "sas/qtgui/MainWindow/MainWindow.py", line 44, in __init__
  File "PyInstaller/loader/pyimod02_importers.py", line 493, in exec_module
  File "sas/qtgui/MainWindow/GuiManager.py", line 35, in <module>
ModuleNotFoundError: No module named 'sas.qtgui.Utilities.Reports'
[93029] Failed to execute script 'sasview' due to unhandled exception: No module named 'sas.qtgui.Utilities.Reports'
[93029] Traceback:
Traceback (most recent call last):
  File "sasview.py", line 15, in <module>
  File "sas/qtgui/MainWindow/MainWindow.py", line 113, in run_sasview
  File "sas/qtgui/MainWindow/MainWindow.py", line 44, in __init__
  File "PyInstaller/loader/pyimod02_importers.py", line 493, in exec_module
  File "sas/qtgui/MainWindow/GuiManager.py", line 35, in <module>
ModuleNotFoundError: No module named 'sas.qtgui.Utilities.Reports'

Does it work on Windows?
I guess there maybe some path missing in pyinstaller file.

**Comment by lucas-wilkins** [comment]
@wpotrzebowski I think @krzywon has fixed this in the as-yet-unmerged #2139

**Comment by lucas-wilkins** [comment]
@wpotrzebowski Try it now?

File: src/sas/qtgui/Perspectives/Corfunc/CorfuncCanvas.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+    @data.setter
+    def data(self, target_data: Optional[Union[Data1DType, Iterable[Data1DType]]]):
+        # I'm not 100% sure this is good practice, but it will make things cleaner in the short term
+        if target_data is None:
+            self._data = None
+        elif isinstance(target_data, Data1D):
+            self._data = [target_data]
+        else:
+            self._data = list(target_data)
```
Comment (rozyczko) [inline src/sas/qtgui/Perspectives/Corfunc/CorfuncCanvas.py:61]: what is the difference between using a `list` function all and using literal `[]`? I thought they would both just return a list?

File: src/sas/qtgui/Perspectives/Corfunc/CorfuncPerspective.py [hunk 2]
```diff
@@ -263,16 +271,16 @@ def transform(self):
...
 
         def updatefn(msg):
             """Report progress of transformation."""
             self.communicate.statusBarUpdateSignal.emit(msg)
 
-        def completefn(transforms):
+        def completefn(transformed_data: Tuple[Data1D, Data1D, Data1D]):
             """Extract the values from the transforms and plot"""
-            self.trigger.emit(transforms)
+            self.trigger.emit(TransformedData(*transformed_data)) # TODO: Make this return more structured data earlier
```
Comment (rozyczko) [inline src/sas/qtgui/Perspectives/Corfunc/CorfuncPerspective.py:283]: `TransformedData` should probably have a constructor taking the tuple

File: src/sas/qtgui/Perspectives/Corfunc/CorfuncPerspective.py [hunk 3]
```diff
@@ -313,28 +328,27 @@ def extract(self):
...
+    #         plot_to_add.symbol = 'Line'
+    #         plot_to_add._xaxis = "x"
+    #         plot_to_add._xunit = "A"
+    #         plot_to_add._yaxis = "\Gamma"
+    #         if i < len(titles):
+    #             title = titles[i]
+    #             plot_to_add.name = titles[i]
+    #         GuiUtils.updateModelItemWithPlot(self._model_item, plot_to_add, title)
+    #         #self.axes.set_xlim(min(data1.x), max(data1.x) / 4)
+
```
Comment (rozyczko) [inline src/sas/qtgui/Perspectives/Corfunc/CorfuncPerspective.py:351]: If unused - please delete. If required for future additions - please add a comment

File: src/sas/qtgui/Perspectives/Corfunc/RealSpaceCanvas.py [hunk 4]
```diff
@@ -44,12 +23,13 @@ def draw_data(self):
...
         self.fig.tight_layout()
 
-        if self.data:
-            data1, data3, data_idf = self.data
-            self.axes.plot(data1.x, data1.y, label="1D Correlation")
-            self.axes.plot(data3.x, data3.y, label="3D Correlation")
-            self.axes.set_xlim(0, max(data1.x) / 4)
-            self.legend = self.axes.legend()
+        if self.data is not None:
+            if len(self.data) == 2:
```
Comment (rozyczko) [inline src/sas/qtgui/Perspectives/Corfunc/RealSpaceCanvas.py:27]: can be squeezed into one line

File: src/sas/sascalc/data_util/calcthread.py [hunk 5]
```diff
@@ -119,30 +111,33 @@ def OnCalcComplete(self,event):
...
+    def queue(self, *args, **kwargs):
         """Add a work unit to the end of the queue.  See the compute()
         method for details of the arguments to the work unit."""
         self._lock.acquire()
         self._queue.append((args, kwargs))
         # Cannot do start_new_thread call within the lock
         self._lock.release()
         if not self._running:
-            self._time_for_update = clock() + 1e6
+            self._time_for_update = perf_counter() + 1e6
```
Comment (rozyczko) [inline src/sas/sascalc/data_util/calcthread.py:140]: Good idea replacing with `perf_counter` - less block-prone.

**Review by rozyczko — APPROVED** [review]
Looks good, minor issues raised

**Comment by wpotrzebowski** [comment]
Indeed, it seems to work now (from installer)

**Comment by wpotrzebowski** [comment]
I've tried it on two files that I think it used to work before (see attached screen shot). 
Also Corfunc window on laptop screen is too big (and cannot be minimized to fit available space). 
Screenshot 2022-08-16 at 11 31 45

`11:25:29 - INFO:  --- SasView session started, version 5.0.5, 2022 ---
11:25:29 - INFO: Python: 3.8.13 (default, Jul 22 2022, 10:40:33) 
[Clang 12.0.0 (clang-1200.0.32.29)]
11:30:49 - ERROR: 'list' object has no attribute 'x'
Traceback (most recent call last):
  File "sas/qtgui/MainWindow/DataExplorer.py", line 785, in sendData
  File "sas/qtgui/Perspectives/Corfunc/CorfuncPerspective.py", line 469, in setData
  File "sas/qtgui/Perspectives/Corfunc/CorfuncCanvas.py", line 63, in data
  File "sas/qtgui/Perspectives/Corfunc/QSpaceCanvas.py", line 78, in draw_data
AttributeError: 'list' object has no attribute 'x'
11:31:22 - ERROR: 'method' object is not connected
Traceback (most recent call last):
  File "sas/qtgui/MainWindow/DataExplorer.py", line 785, in sendData
  File "sas/qtgui/Perspectives/Corfunc/CorfuncPerspective.py", line 454, in setData
TypeError: 'method' object is not connected`

**Comment by rozyczko** [comment]
Same behaviour on Windows (on sending a file to the corfunc perspective)

```
Traceback (most recent call last):
  File "sas\qtgui\MainWindow\DataExplorer.py", line 785, in sendData
  File "sas\qtgui\Perspectives\Corfunc\CorfuncPerspective.py", line 469, in setData
  File "sas\qtgui\Perspectives\Corfunc\CorfuncCanvas.py", line 63, in data
  File "sas\qtgui\Perspectives\Corfunc\QSpaceCanvas.py", line 78, in draw_data
AttributeError: 'list' object has no attribute 'x'
```

File: src/sas/qtgui/Perspectives/Corfunc/CorfuncPerspective.py (hunk 2 above)
Comment (lucas-wilkins) [reply src/sas/qtgui/Perspectives/Corfunc/CorfuncPerspective.py:283]: It's a namedtuple, really the line above it should be returning it, but all the Data1D issues prevent this.

File: src/sas/qtgui/Perspectives/Corfunc/CorfuncCanvas.py (hunk 1 above)
Comment (lucas-wilkins) [reply src/sas/qtgui/Perspectives/Corfunc/CorfuncCanvas.py:61]: list( [a,b,c] ) = [a,b,c]
[ [a,b,c] ] = [[a,b,c]]

File: src/sas/qtgui/Perspectives/Corfunc/CorfuncCanvas.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+from typing import Optional, Union, List, Iterable, TYPE_CHECKING
+
+from matplotlib.backends.backend_qt5agg import FigureCanvasQTAgg as FigureCanvas
+from matplotlib.figure import Figure
+
+# TODO: Make not crazy
+from sas.qtgui.Plotting.PlotterData import Data1D as Data1DVersion1
+from sas.sascalc.dataloader.data_info import Data1D as Data1DVersion2
+from sasmodels.data import Data1D as Data1DVersion3
+Data1D = (Data1DVersion1, Data1DVersion2, Data1DVersion3)
```
[the remark this answers is not in the document]
Comment (lucas-wilkins) [reply src/sas/qtgui/Perspectives/Corfunc/CorfuncCanvas.py:14]: Yeah, you'd think! Really need to work on Data1D, mostly waiting on @krzywon's changes to do this.

File: src/sas/qtgui/Perspectives/Corfunc/CorfuncCanvas.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
+from __future__ import annotations
+
+from abc import ABC, ABCMeta, abstractmethod
```
[the remark this answers is not in the document]
Comment (lucas-wilkins) [reply src/sas/qtgui/Perspectives/Corfunc/CorfuncCanvas.py:3]: oops

File: src/sas/qtgui/Perspectives/Corfunc/CorfuncPerspective.py (hunk 3 above)
Comment (lucas-wilkins) [reply src/sas/qtgui/Perspectives/Corfunc/CorfuncPerspective.py:351]: done

**Comment by lucas-wilkins** [comment]
@rozyczko @wpotrzebowski Ah, this was broken by the merge from main to fix the installer.

**Review by smk78 — COMMENTED** [review]
Functionality test on W10/x64.
Does what it advertises. Just one minor bug: the legend for the IDF does not appear.
image
The Log Explorer says:
```
 WARNING: No handles with labels found to put in legend.
```

**Comment by rozyczko** [comment]
> @rozyczko @wpotrzebowski Ah, this was broken by the merge from main to fix the installer.

and it works now!

**Comment by butlerpd** [comment]
@lucas-wilkins will merge once fixing the Mac OS window problem

File: src/sas/sascalc/data_util/calcthread.py [hunk 8]
```diff
@@ -119,30 +111,33 @@ def OnCalcComplete(self,event):
...
 
-    def __init__(self, completefn=None, updatefn=None,
-                 yieldtime=0.01, worktime=0.01,
-                 exception_handler=None):
-        """Prepare the calculator"""
-        self.yieldtime     = yieldtime
-        self.worktime      = worktime
-        self.completefn    = completefn
-        self.updatefn      = updatefn
+    def __init__(self,
```
Comment (krzywon) [inline src/sas/sascalc/data_util/calcthread.py:114]: This API change inadvertently breaks fitting. `calcthread.py` is also used by `FittingWidget.py` and it passes arguments by keyword on line 2888.

**Review by krzywon — CHANGES_REQUESTED** [review]
Functionally, this is a huge improvement to the Corfunc perspective. I am still working on a full code review but wanted to put this change request this isn't merged before I finish that review.

**Review by krzywon — COMMENTED** [review]
Here are some further review comments. The only other thing that needs to be fixed is the IDF y-axis label. Treat my other comments as suggestions.

File: src/sas/sascalc/data_util/calcthread.py (hunk 8 above)
Comment (lucas-wilkins) [reply src/sas/sascalc/data_util/calcthread.py:114]: Ah, thought I'd caught them all. This is why you don't pass functions around.

File: src/sas/qtgui/Perspectives/Corfunc/CorfuncCanvas.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+from matplotlib.backends.backend_qt5agg import FigureCanvasQTAgg as FigureCanvas
+from matplotlib.figure import Figure
+
+# TODO: Make not crazy
+from sas.qtgui.Plotting.PlotterData import Data1D as Data1DVersion1
+from sas.sascalc.dataloader.data_info import Data1D as Data1DVersion2
+from sasmodels.data import Data1D as Data1DVersion3
+Data1D = (Data1DVersion1, Data1DVersion2, Data1DVersion3)
+Data1DType = Union[Data1DVersion1, Data1DVersion2, Data1DVersion3]
```
[the remark this answers is not in the document]
Comment (lucas-wilkins) [reply src/sas/qtgui/Perspectives/Corfunc/CorfuncCanvas.py:15]: I was getting type errors, but maybe I messed up somehow.

File: src/sas/qtgui/Perspectives/Corfunc/IDFCanvas.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+        are all draw in on the plot in linear cooredinates."""
+
+
+        self.fig.clf()
+
+        self.axes = self.fig.add_subplot(111)
+        self.axes.set_xscale("linear")
+        self.axes.set_yscale("linear")
+        self.axes.set_xlabel("Z [$\AA$]")
+        self.axes.set_ylabel("IDF")
```
[the remark this answers is not in the document]
Comment (lucas-wilkins) [reply src/sas/qtgui/Perspectives/Corfunc/IDFCanvas.py:23]: RealSpaceCanvas: When I'm sure about what type it needs, I'll type it.

[the remark this answers is not in the document]
Comment (lucas-wilkins) [reply src/sas/qtgui/Perspectives/Corfunc/IDFCanvas.py:23]: IDFCanvas: odd indeed, was planning on investigating before pushing

File: src/sas/qtgui/Perspectives/Corfunc/RealSpaceCanvas.py (hunk 4 above)
Comment (lucas-wilkins) [reply src/sas/qtgui/Perspectives/Corfunc/RealSpaceCanvas.py:27]: Its a suggestion in the same way that a police officer might suggest that you've had too much to drink and that its time to go home.

File: src/sas/sascalc/data_util/calcthread.py (hunk 8 above)
Comment (lucas-wilkins) [reply src/sas/sascalc/data_util/calcthread.py:114]: I've just reverted the changes.

File: src/sas/qtgui/Perspectives/Corfunc/CorfuncCanvas.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (lucas-wilkins) [reply src/sas/qtgui/Perspectives/Corfunc/CorfuncCanvas.py:15]: Sorted

File: src/sas/qtgui/Perspectives/Corfunc/IDFCanvas.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (lucas-wilkins) [reply src/sas/qtgui/Perspectives/Corfunc/IDFCanvas.py:23]: Yeah, its there on my machine. Just matplotlib draws it outside the window if its small.

**Review by krzywon — APPROVED** [review]

**Comment by lucas-wilkins** [comment]
@wpotrzebowski @smk78  Better?

**Comment by wpotrzebowski** [comment]
Window scalling works fine now. 

Two potential issues that arised while testing
1. Label and legend on IDF plot missing (as mentioned on the call on Tuesday)
2. When trying to generate report I get the following error: 
```
8:52:20 - INFO:  --- SasView session started, version 5.0.5, 2022 ---
08:52:20 - INFO: Python: 3.8.13 (default, Jul 22 2022, 10:40:33) 
[Clang 12.0.0 (clang-1200.0.32.29)]
08:52:47 - WARNING: Background value results in negative intensities (1.051 > 0.44883)
08:54:14 - ERROR: Traceback (most recent call last):
  File "sas/qtgui/MainWindow/GuiManager.py", line 840, in actionReport
  File "sas/qtgui/Perspectives/Corfunc/CorfuncPerspective.py", line 788, in getReport
  File "sas/qtgui/Utilities/Reports/reports.py", line 106, in __init__
  File "importlib/resources.py", line 169, in read_text
  File "importlib/resources.py", line 125, in open_text
  File "PyInstaller/loader/pyimod02_importers.py", line 539, in open_resource
  File "pathlib.py", line 1222, in open
  File "pathlib.py", line 1078, in _opener
FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/q3/5vy4nct93wg03ynnn836g49m0000gq/T/AppTranslocation/58FB7774-5D37-4038-9D15-6BD37D4AB3DB/d/SasView5.app/Contents/MacOS/sas/qtgui/Utilities/Reports/report_style.css' 
```

**Comment by lucas-wilkins** [comment]
> Label and legend on IDF plot missing (as mentioned on the call on Tuesday)

Legend was deleted, as there's only one curve on the IDF plot.

Axis label is not missing, it's just a long way to the left - it shows on my screen, which is perhaps wider than yours - if you have a sensible fix for it let me know.

> When trying to generate report I get the following error:

Ooer, I'll look at that,, think its probably about packaging the CSS. I thought @krzywon's fix should have sorted that.

**Comment by rozyczko** [comment]
> > When trying to generate report I get the following error:
> 
> Ooer, I'll look at that,, think its probably about packaging the CSS. I thought @krzywon's fix should have sorted that.

I think the CSS file needs to be put somewhere more central than 
`SasView\\sas\\qtgui\\Utilities\\Reports\\report_style.css` 
We probably don't want to recreate this path in the installed app

**Review by smk78 — COMMENTED** [review]
Functionality test on W10/x64 using a 1920x1080 monitor.
When I first brought up the Corfunc perspective the bottom of the perspective was somewhere in the midst of the Lamellar Parameters dialog and I had to stretch the perspective downwards, which I was a little surprised about as there was plenty of screen real estate.
I note the (missing) IDF legend has now been removed completely. Which is fine. But as others have noted, the y-axis label is not initially displaying correctly:
image
However, stretching the perspective window to the right fixes this.
Saving extrapolated & transformed data to file worked fine.
Generating a report worked fine, and looked fine in the report dialog window, but the actual output is real ugly and needs looking at.
[corfunc_report.pdf](https://github.com/SasView/sasview/files/9381361/corfunc_report.pdf)
At some point in what I was doing this appeared in the Console, possibly when I was 'reporting':
```
WARNING: getSize: Not a float '100%'
```

**Comment by lucas-wilkins** [comment]
> When I first brought up the Corfunc perspective the bottom of the perspective was somewhere in the midst of the Lamellar Parameters dialog and I had to stretch the perspective downwards, which I was a little surprised about as there was plenty of screen real estate.

Yeah, it's either that or no scrolling. Unless you can think of a way of coercing the window to a good size.

**Comment by lucas-wilkins** [comment]
CSS file should now be included in the installers

**Comment by lucas-wilkins** [comment]
> WARNING: getSize: Not a float '100%'

That's the html to pdf converter not liking how the width of the images is specified.

**Review by lucas-wilkins — COMMENTED** [review]
@wpotrzebowski this should fix the report issue you had

**Review by smk78 — APPROVED** [review]
Testing latest changes on W10/x64. Plots are now separated in the report.

**Comment by lucas-wilkins** [comment]
SasView can find the css file now, but doesn't have permission to read it for some reason.

**Comment by lucas-wilkins** [comment]
Woooooo, finally, reports bugs all squashed!

**Pull request merged**